Libraries/rocFFT/callback/main.cpp

```diff
@@ -0,0 +1,175 @@
+
+#include <iostream>
+#include <math.h>
+#include <random>
+#include <vector>
+
+// example of using load/store callbacks with rocfft
+
+struct load_callback_data
+{
+    double2* filter;
+    double   scale;
+};
+
+__device__ double2 load_callback(double2* input,
+                                 size_t   offset,
+                                 void*    callback_data,
+                                 void* /*sharedMem*/)
+{
+    auto data = static_cast<load_callback_data*>(callback_data);
+
+    // multiply each element by filter element and scale
+    return hipCmul(hipCmul(input[offset], data->filter[offset]),
+                   make_hipDoubleComplex(data->scale, data->scale));
+}
+
+// Can not give __device__ function to HIP_SYMBOL
+__device__ auto load_callback_dev = load_callback;
+
+int main()
+{
+    constexpr size_t N = 8;
+
+    std::vector<double2> input(N), callback_filter(N);
+
+    // Initialize data and filter
+    std::random_device                     rd;
+    std::default_random_engine             gen(rd());
+    std::uniform_real_distribution<double> distribution(0.0, 1.0);
+
+    for(size_t i = 0; i < N; i++)
+    {
+        input[i].x           = i;
+        input[i].y           = i;
+        callback_filter[i].x = distribution(gen);
+    }
+
+    // Rocfft gpu compute
+    ROCFFT_CHECK(rocfft_setup());
+
+    const size_t Nbytes = N * sizeof(double2);
+
+    // Create HIP device object.
+    double2 *data_dev, *callback_filter_dev;
+
+    // Create buffers
+    HIP_CHECK(hipMalloc(&data_dev, Nbytes));
+    HIP_CHECK(hipMalloc(&callback_filter_dev, Nbytes));
+
+    // Copy data to device
+    HIP_CHECK(hipMemcpy(data_dev, input.data(), Nbytes, hipMemcpyHostToDevice));
+    HIP_CHECK(
+        hipMemcpy(callback_filter_dev, callback_filter.data(), Nbytes, hipMemcpyHostToDevice));
+
+    // Create plan
+    rocfft_plan plan = nullptr;
+    ROCFFT_CHECK(rocfft_plan_create(&plan,
+                                    rocfft_placement_inplace,
+                                    rocfft_transform_type_complex_forward,
+                                    rocfft_precision_double,
+                                    1,
+                                    &N,
+                                    1,
+                                    nullptr));
+
+    // Check if the plan requires a work buffer
+    size_t work_buf_size = 0;
+    ROCFFT_CHECK(rocfft_plan_get_work_buffer_size(plan, &work_buf_size));
+    rocfft_execution_info info = nullptr;
+    ROCFFT_CHECK(rocfft_execution_info_create(&info));
+
+    void* work_buf = nullptr;
+    if(work_buf_size)
+    {
+        HIP_CHECK(hipMalloc(&work_buf, work_buf_size));
+        ROCFFT_CHECK(rocfft_execution_info_set_work_buffer(info, work_buf, work_buf_size));
+    }
+
+    // Prepare callback
+    load_callback_data callback_data_host;
+    callback_data_host.filter = callback_filter_dev;
+    callback_data_host.scale  = 1.0 / static_cast<double>(N);
```

Comment:
rocFFT has an explicit API to perform result scaling:  https://rocm.docs.amd.com/projects/rocFFT/en/latest/how-to/working-with-rocfft.html#result-scaling

This API is expected to perform better than callbacks, though I realize that the rocFFT repository does not currently have an example to demonstrate its usage.  The explicit API is quicker because the compiler is able to understand and optimize the extra scaling multiplication - with callback functions the runtime only receives an opaque function pointer and the compiler cannot optimize the code as well.

AFAICT result scaling is the most common use case for callback functions in the API.  I think it would make sense to have an example demonstrating use of the explicit result scaling API, and then this callback example can be repurposed for a different operation.

I don't know which operation would be best though - aside from result scaling, I'm unaware of a commonly-used operation that people would want to do in a callback function.

Reply:
Oh I see, we hadn't noticed this functionality before. Perhaps, given that the callback applies a filter and a scaling factor, we can keep the filtering and just use the scaling-specific API you mentioned for the scaling part. I think that still would make sense as a use-case for callbacks, and then we also get to showcase roc/hipFFT's scaling API.

Reply:
I guess that's reasonable, since rocFFT has no API for filtering currently.  The only issue is that we wouldn't really see the performance benefit of using the result scaling API, since callbacks are still in  use.  We hope to have a better way to express filtering (and other spectral operations) eventually, but it'll take a while to get there.

Reply:
Forgot to answer here, the example should be updated already!